### PR TITLE
Truncate page hit title

### LINKED ISSUE #7367: Error inserting page hit because url title is too long
[//]: # ( Invisible comment: 
IIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIII
Before you create the issue:
IIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIII
Search for similar report among other reported issues.
Learn how to troubleshoot at https://www.mautic.org/docs/en/tips/troubleshooting.html
Use drag&drop to attach images or other files )

## Bug Description
Column `hit.url_title` is a string (`varchar(255)`), but title is unlimited.

| Q   | A
| --- | ---
| Mautic version | 2.15.1
| PHP version | 7.2
| Browser | 

### Steps to reproduce
1. Add mautic tracking to a page with long title
2. Visit it
 
### Log errors

> (Doctrine\\DBAL\\Exception\\DriverException(code: 0): An exception occurred while executing 'UPDATE page_hits SET country = ?, region = ?, city = ?, isp = ?, organization = ?, referer = ?, url = ?, url_title = ?, user_agent = ?, page_language = ?, browser_languages = ?, lead_id = ? WHERE id = ?' with params [\"Brazil\", \"Bahia\", \"Salvador\", \"\", \"\", \"www.example.com\", \"https:\\/\\/www.example.com\", \"Long Title Long Title Long Title Long Title Long Title Long Title Long Title Long Title Long Title Long Title Long Title Long Title Long Title Long Title Long Title Long Title Long Title Long Title Long Title Long Title Long Title Long Title Long Title Long Title Long Title Long Title\", \"Mozilla\\/5.0 (Macintosh; Intel Mac OS X 10_14_2) AppleWebKit\\/605.1.15 (KHTML, like Gecko) Version\\/12.0.2 Safari\\/605.1.15\", \"pt-BR\", \"a:1:{i:0;s:5:\\\"pt-br\\\";}\", 117777, 278905]:\n\nSQLSTATE[22001]: String data, right truncated: 1406 Data too long for column 'url_title' at row 1 at /var/www/html/vendor/doctrine/dbal/lib/Doctrine/DBAL/Driver/AbstractMySQLDriver.php:115, Doctrine\\DBAL\\Driver\\PDOException(code: 22001): SQLSTATE[22001]: String data, right truncated: 1406 Data too long for column 'url_title' at row 1 at /var/www/html/vendor/doctrine/dbal/lib/Doctrine/DBAL/Driver/PDOStatement.php:107, PDOException(code: 22001): SQLSTATE[22001]: String data, right truncated: 1406 Data too long for column 'url_title' at row 1 at /var/www/html/vendor/doctrine/dbal/lib/Doctrine/DBAL/Driver/PDOStatement.php:105)

[//]: # ( Invisible comment:
Please check for related errors in the latest log file in [mautic root]/app/log/ and/or the web server's logs and post them here. Be sure to remove sensitive information if applicable. )

### PR DESCRIPTION
**Pull request by kuzmany**
| Q                                      | A
| -------------------------------------- | ---
| Bug fix? (use the a.b branch)          | [ ]
| New feature/enhancement? (use the a.x branch)      | [ ]
| Deprecations?                          | [ ]
| BC breaks? (use the c.x branch)        | [ ]
| Automated tests included?              | [ ] 
| Related user documentation PR URL      | mautic/mautic-documentation#... 
| Related developer documentation PR URL | mautic/developer-documentation#... 
| Issue(s) addressed                     | Fixes #... 

#### Description:

If you use page with tilte longer like 191 chars, then tracking pixel is not triggered. This PR fixed it

#### Steps to test this PR:

1. Open this PR on Gitpod or pull down for testing locally (see docs on testing PRs [here](https://contribute.mautic.org/contributing-to-mautic/tester))
2. Create page with title longer like 191 chars like this

```html
<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="UTF-8">
  <title>Title longer than 191 chars Title longer than 191 chars Title longer than 191 chars Title longer than 191 chars Title longer than 191 chars Title longer than 191 chars Title longer than 191 chars Title longer than 191 chars Title longer than 191 chars Title longer than 191 chars Title longer than 191 chars Title longer than 191 chars Title longer than 191 chars </title>

</head>
<body>

</html>
```
3. Put tracking pixel on the page
4. Open the page
5. Tracking pixel is not triggered
6. After PR tracking pixel is triggered with title of page truncated to 191 chars

<a href="https://gitpod.io/#https://github.com/mautic/mautic/pull/10709"></a>

### CODE REVIEW

**Comment by mmarton** [comment]
Hi!

I've tested this with title lengths 190 and 2000 characters (linux mint, firefox 95, chrome 96)

I could not confirm the original issue. The tracking worked for me with any title length.

sidenotes:

- mautic throws an exception if you try to save a page with title longer than 190 (could use a length validation, is this PR a good candidate for that or should I open a new issue?) 
> 500 Internal Server Error - An exception occurred while executing 'UPDATE pages SET date_modified = ?, checked_out = ?, checked_out_by = ?, title = ?, revision = ? WHERE id = ?' with params ["2021-12-29 14:23:35", null, null, "123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012", 4, 1]: SQLSTATE[22001]: String data, right truncated: 1406 Data too long for column 'title' at row 1 

- Could't use gitpod for testing because the pagebuilder throw a cors js error, so I tested locally, is there a way to make it work?

- this is my first review, please correct me if I made any mistakes. Thankyou

**Comment by mmarton** [comment]
Hi!

I've tested this with title lengths 190 and 2000 characters (linux mint, firefox 95, chrome 96)

I could not confirm the original issue. The tracking worked for me with any title length.

sidenotes:

    mautic throws an exception if you try to save a page with title longer than 190 (could use a length validation, is this PR a good candidate for that or should I open a new issue?)

    500 Internal Server Error - An exception occurred while executing 'UPDATE pages SET date_modified = ?, checked_out = ?, checked_out_by = ?, title = ?, revision = ? WHERE id = ?' with params ["2021-12-29 14:23:35", null, null, "123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012", 4, 1]: SQLSTATE[22001]: String data, right truncated: 1406 Data too long for column 'title' at row 1

    Could't use gitpod for testing because the pagebuilder throw a cors js error, so I tested locally, is there a way to make it work?

    this is my first review, please correct me if I made any mistakes. Thankyou

**Review by mmarton — COMMENTED** [review]
Hi!

I've tested this with title lengths 190 and 2000 characters (linux mint, firefox 95, chrome 96)

I could not confirm the original issue. The tracking worked for me with any title length.

sidenotes:

 - mautic throws an exception if you try to save a page with title longer than 190 (could use a length validation, is this PR a good candidate for that or should I open a new issue?)
> 500 Internal Server Error - An exception occurred while executing 'UPDATE pages SET date_modified = ?, checked_out = ?, checked_out_by = ?, title = ?, revision = ? WHERE id = ?' with params ["2021-12-29 14:23:35", null, null, "123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012", 4, 1]: SQLSTATE[22001]: String data, right truncated: 1406 Data too long for column 'title' at row 1

- Could't use gitpod for testing because the pagebuilder throw a cors js error, so I tested locally, is there a way to make it work?

- this is my first review, please correct me if I made any mistakes. Thankyou

File: app/bundles/PageBundle/Entity/Hit.php [hunk 1]
```diff
@@ -558,6 +558,7 @@ public function getUrl()
      */
     public function setUrlTitle($urlTitle)
     {
+        $urlTitle       = mb_strlen($urlTitle) <= 191 ? $urlTitle : substr($urlTitle, 0, 191);
```
Comment (escopecz) [inline app/bundles/PageBundle/Entity/Hit.php]: Let's use the multibyte functions
```suggestion
        $urlTitle       = mb_strlen($urlTitle) <= 191 ? $urlTitle : mb_substr($urlTitle, 0, 191);
```

File: app/bundles/PageBundle/Tests/Entity/HitTest.php [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+{
+    /**
+     * @dataProvider setUrlTitle
+     */
+    public function testSetUrlTitle(string $urlTitle, int $expected): void
+    {
+        $hit = new Hit();
+        $hit->setUrlTitle($urlTitle);
+
+        Assert::assertEquals($expected, strlen($hit->getUrlTitle()));
```
Comment (escopecz) [inline app/bundles/PageBundle/Tests/Entity/HitTest.php]:
```suggestion
        Assert::assertEquals($expected, mb_substr($hit->getUrlTitle()));
```

**Review by escopecz — CHANGES_REQUESTED** [review]
Thanks Zdeno! 2 small suggestions.

**Comment by escopecz** [comment]
@mmarton can you please check that you were able to store a page title longer than 255 characters in the page_hits table before this change? I found an issue (https://github.com/mautic/mautic/issues/7367) reporting the same problem.

And yes, the error you found should be fixed in another PR. It's not directly related to this one.

**Review by escopecz — DISMISSED** [review]
The code make sense, I can see the `page_hits.url_title` column is type of `VARCHAR(191)`, the test confirms it works as expected 👍

File: app/bundles/PageBundle/Tests/Entity/HitTest.php [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+{
+    /**
+     * @dataProvider setUrlTitle
+     */
+    public function testSetUrlTitle(string $urlTitle, int $expected): void
+    {
+        $hit = new Hit();
+        $hit->setUrlTitle($urlTitle);
+
+        Assert::assertEquals($expected, mb_substr($hit->getUrlTitle()));
```
Comment (escopecz) [inline app/bundles/PageBundle/Tests/Entity/HitTest.php]: OOps, I meant
```suggestion
        Assert::assertEquals($expected, mb_strlen($hit->getUrlTitle()));
```

**Review by escopecz — DISMISSED** [review]
Re-approving

**Review by mabumusa1 — APPROVED** [review]
I did a code review, I think it works and solve a problem for long URL titles

**Pull request merged**